README.rst
kotti_audit
***********

Add an audit log to your Kotti website.

Features
===========
- Content audit log for editors to find content based on creation or modification.
- Each content node has an audit log that display all children of that node.
- Content filtering and sorthing by title, content type, container title, modification_date and creation_date
- Content search

Note, this package adds the ***Audit Log*** menu entry to the ***Actions*** menu.
The Audit Log menu entry will take you the content respective Audit view, e.g.
http://localhost/audit-log or http://localhost/example-content/audit-log

Example
=========

.. image:: https://user-images.githubusercontent.com/1830494/77401539-19d4c180-6d7b-11ea-88db-fb1e5be3cd4f.png
  :width: 100%
  :alt: Audit Log for Kotti


|pypi|_
|downloads_month|_
|license|_
|build_status_stable|_

.. |pypi| image:: https://img.shields.io/pypi/v/kotti_audit.svg?style=flat-square
.. _pypi: https://pypi.python.org/pypi/kotti_audit/

.. |downloads_month| image:: https://img.shields.io/pypi/dm/kotti_audit.svg?style=flat-square
.. _downloads_month: https://pypi.python.org/pypi/kotti_audit/

.. |license| image:: https://img.shields.io/pypi/l/kotti_audit.svg?style=flat-square
.. _license: http://www.repoze.org/LICENSE.txt

.. |build_status_stable| image:: https://img.shields.io/travis/jamaicanDevelopers/kotti_audit/production.svg?style=flat-square
.. _build_status_stable: http://travis-ci.org/jamaicanDevelopers/kotti_audit

`Find out more about Kotti`_

Development happens at https://github.com/jamaicanDevelopers/kotti_audit

.. _Find out more about Kotti: http://pypi.python.org/pypi/Kotti

Setup
=====

To enable the extension in your Kotti site, activate the configurator::

    kotti.configurators =
        kotti_audit.kotti_configure

Database upgrade
================

If you are upgrading from a previous version you might have to migrate your
database.  The migration is performed with `alembic`_ and Kotti's console script
``kotti-migrate``. To migrate, run
``kotti-migrate upgrade --scripts=kotti_audit:alembic``.

For integration of alembic in your environment please refer to the
`alembic documentation`_. If you have problems with the upgrade,
please create a new issue in the `tracker`_.

Development
===========

|build_status_master|_

.. |build_status_master| image:: https://img.shields.io/travis/jamaicanDevelopers/kotti_audit/master.svg?style=flat-square
.. _build_status_master: http://travis-ci.org/jamaicanDevelopers/kotti_audit

Contributions to kotti_audit are highly welcome.
Just clone its `Github repository`_ and submit your contributions as pull requests.

.. _alembic: http://pypi.python.org/pypi/alembic
.. _alembic documentation: https://alembic.readthedocs.io/en/latest/index.html
.. _tracker: https://github.com/jamaicanDevelopers/kotti_audit/issues
.. _Github repository: https://github.com/jamaicanDevelopers/kotti_audit


CHANGES.rst
History
=======

0.0.1 (unreleased)
------------------

- Create package with ``cookiecutter https://github.com/Kotti/kotti-cookiecutter``.


kotti_audit/tests/conftest.py
# -*- coding: utf-8 -*-

"""
Created on 2020-03-21
:author: Oshane Bailey (b4.oshany@gmail.com)
"""

pytest_plugins = "kotti"

from pytest import fixture


@fixture(scope='session')
def custom_settings():
    import kotti_audit.resources
    kotti_audit.resources  # make pyflakes happy
    return {
        'kotti.configurators': 'kotti_tinymce.kotti_configure '
                               'kotti_audit.kotti_configure'}


kotti_audit/tests/test_view.py
# -*- coding: utf-8 -*-

"""
Created on 2020-03-21
:author: Oshane Bailey (b4.oshany@gmail.com)
"""

from pytest import fixture


@fixture
def dummy_content(root):

    from kotti_audit.resources import CustomContent

    root['cc'] = cc = CustomContent(
        title=u'My content',
        description=u'My very custom content is custom',
        custom_attribute='Lorem ipsum'
    )

    return cc


def test_view(dummy_content, dummy_request):

    from kotti_audit.views.view import CustomContentViews

    views = CustomContentViews(dummy_content, dummy_request)

    default = views.default_view()
    assert 'foo' in default

    alternative = views.alternative_view()
    assert alternative['foo'] == u'bar'


kotti_audit/tests/test_resources.py
# -*- coding: utf-8 -*-

"""
Created on 2020-03-21
:author: Oshane Bailey (b4.oshany@gmail.com)
"""

from pytest import raises


def test_model(root, db_session):
    from kotti_audit.resources import CustomContent

    cc = CustomContent()
    assert cc.custom_attribute is None

    cc = CustomContent(custom_attribute=u'Foo')
    assert cc.custom_attribute == u'Foo'

    root['cc'] = cc = CustomContent()
    assert cc.name == 'cc'

    with raises(TypeError):
        cc = CustomContent(doesnotexist=u'Foo')


i18n.sh
#!/bin/bash
# Usage:
#     Initial catalog creation (lang is the language identifier):
#         ./i18n.sh lang
#     Updating translation and compile catalog:
#         ./i18n.sh

# configuration
DOMAIN="kotti_audit"
SEARCH_PATH="kotti_audit"
LOCALES_PATH="kotti_audit/locale"
# end configuration

# create locales folder if not exists
if [ ! -d "$LOCALES_PATH" ]; then
    echo "Locales directory not exists, create"
    mkdir -p "$LOCALES_PATH"
fi

# create pot if not exists
if [ ! -f "$LOCALES_PATH"/$DOMAIN.pot ]; then
    echo "Create pot file"
    touch "$LOCALES_PATH"/$DOMAIN.pot
fi

# no arguments, extract and update
if [ $# -eq 0 ]; then
    echo "Extract messages"
    pot-create "$SEARCH_PATH" -d $DOMAIN -o "$LOCALES_PATH"/$DOMAIN.pot

    echo "Update translations"
    for po in "$LOCALES_PATH"/*/LC_MESSAGES/$DOMAIN.po; do
        msgmerge --no-wrap -o "$po" "$po" "$LOCALES_PATH"/$DOMAIN.pot
    done

    echo "Compile message catalogs"
    for po in "$LOCALES_PATH"/*/LC_MESSAGES/*.po; do
        echo "Compiling file $po..."
        msgfmt --statistics -o "${po%.*}.mo" "$po"
    done

# first argument represents language identifier, create catalog
else
    cd "$LOCALES_PATH"
    mkdir -p $1/LC_MESSAGES
    msginit -i $DOMAIN.pot -o $1/LC_MESSAGES/$DOMAIN.po -l $1
fi


kotti_audit/__init__.py
# -*- coding: utf-8 -*-

"""
Created on 2020-03-21
:author: Oshane Bailey (b4.oshany@gmail.com)
"""

from kotti.resources import File
from pyramid.i18n import TranslationStringFactory
from kotti.util import Link
from kotti.resources import default_actions
from kotti.resources import Content
from kotti.views.site_setup import CONTROL_PANEL_LINKS as KOTTI_USER_NAV_LINKS


_ = TranslationStringFactory('kotti_audit')

default_actions.insert(0, Link('audit-log', title=_(u'Audit Log')))
KOTTI_USER_NAV_LINKS.append(
    Link('audit-log', title=_(u'Audit Log'))
)

def kotti_configure(settings):
    """ Add a line like this to you .ini file::

            kotti.configurators =
                kotti_audit.kotti_configure

        to enable the ``kotti_audit`` add-on.

    :param settings: Kotti configuration dictionary.
    :type settings: dict
    """

    settings['pyramid.includes'] += ' kotti_audit'
    settings['kotti.alembic_dirs'] += ' kotti_audit:alembic'
    try:
        
        from kotti_controlpanel import CONTROL_PANEL_LINKS as CPANEL_PAGE_LINKS
        CPANEL_PAGE_LINKS.append(
            Link('audit-log', title=_(u'Audit Log'))
        )
        if 'kotti_controlpanel.kotti_configure' not in settings[
            'kotti.configurators']:
            settings['kotti.configurators'] += (
                '\nkotti_controlpanel.kotti_configure'
            )
            settings['kotti.base_includes'] += (
                ' kotti_controlpanel'
            )
    except ImportError:
        pass

    settings['kotti.fanstatic.view_needed'] += (
        ' kotti_audit.fanstatic.css_and_js'
    )


def includeme(config):
    """ Don't add this to your ``pyramid_includes``, but add the
    ``kotti_configure`` above to your ``kotti.configurators`` instead.

    :param config: Pyramid configurator object.
    :type config: :class:`pyramid.config.Configurator`
    """

    config.add_translation_dirs('kotti_audit:locale')
    config.add_static_view('static-kotti_audit', 'kotti_audit:static')

    config.scan(__name__)


kotti_audit/views/view.py
# -*- coding: utf-8 -*-

"""
Created on 2020-03-21
:author: Oshane Bailey (b4.oshany@gmail.com)
"""
import json
from pyramid.view import view_config
from pyramid.view import view_defaults
from sqlalchemy import or_
from sqlalchemy.orm import aliased

from kotti_audit import _
from kotti_audit import utils
from kotti_audit.fanstatic import css_and_js
from kotti_audit.views import BaseView
from kotti.resources import Content


@view_defaults(permission='edit', context=Content)
class AuditLogViews(BaseView):
    """ Views for :class:`kotti_audit.resources.CustomContent` """

    @view_config(name='audit-log', permission='view',
                 renderer='kotti_audit:templates/audit-log.pt')
    def default_view(self):
        """ Default view`

        :result: Dictionary needed to render the template.
        :rtype: dict
        """
        css_and_js.need()
        
        content_types = Content.query.distinct(
            Content.type).with_entities(Content.type).all();

        return {
            'type_filters': json.dumps(
                {t.type: t.type.title() for t in content_types})
        }

    @view_config(name="audit-api", permission="admin",
                 renderer="json")
    def audit_api(self):
        """ API view`

        :result: Dictionary needed to render the template.
        :rtype: dict
        """
        limit = int(self.request.params.get("limit", 50))
        offset = int(self.request.params.get("offset", 0))
        sort = self.request.params.get("sort", "modification_date")
        order = self.request.params.get("order", "desc")
        filters = self.request.params.get("filter", None)
        search = self.request.params.get("search", None)

        query = Content.query;

        # Use the current context as the base location of the search,
        # otherwise, use root as the base location if no parent is found.
        parent_id = self.context.id if self.context.__parent__ else 0
        if parent_id != 0:
            query = query.filter(
                Content.path.ilike('{}%'.format(self.context.path)),
                Content.id != self.context.id
            )

        if filters is not None:
            filters = json.loads(filters)
        
            if 'creation_date' in filters:
                creation_date = filters['creation_date']
                query = utils.date_query(query, 'creation_date', creation_date)
        
            if 'modification_date' in filters:
                modification_date = filters['modification_date']
                query = utils.date_query(query, 'modification_date', modification_date)
            
            if 'title' in filters:
                query = query.filter(
                    Content.title.ilike("%{}%".format(filters['title'])))
            
            if 'type' in filters:
                query = query.filter_by(type=filters['type'].lower())
            
            if 'parent' in filters:
                parent_title = filters['parent']
                
                parent_ids = Content.query.filter(
                    Content.title.ilike("%{}%".format(parent_title))
                ).values('id')
                
                query = query.filter(
                    Content.parent_id.in_([r[0] for r in parent_ids])
                )


        if search:
            
            # Remvoe the current filters' value from the search text.
            if filters:
                for v in filters.values():
                    search = search.replace(v, '')
            search = search.strip()
            query = query.filter(
                Content.title.ilike("%{}%".format(search))
            )

        if order == 'asc':
            query = query.order_by(Content.__dict__.get(sort).asc())
        else:
            query = query.order_by(Content.__dict__.get(sort).desc())

        total_query = query.count()
        if offset > 0:
            query = query.offset(offset)
        if limit > 0:
            query = query.limit(limit)

        # import pdb; pdb.set_trace()
        # build bootstrap table response.
        query_json = [
            {
                "title": f.title,
                "parent": f.parent.title if f.parent is not None else '',
                "path": f.path,
                'type': f.type.title(),
                "url": f.path,
                "id": f.id,
                "creation_date": f.creation_date.strftime(
                    "%A %d. %B %Y - %I:%M%p %z ") or '',
                "modification_date": f.modification_date.strftime(
                    "%A %d. %B %Y - %I:%M%p %z ") or ''
            }
            for f in query
        ]
        return {
            "rows": query_json,
            "total": total_query
        }
